src/openmc_nek_driver.cpp

```diff
@@ -346,45 +391,56 @@ void OpenmcNekDriver::update_temperature()
 
         // Set temperature for cell instance
         average_temp /= total_vol;
+        Expects(average_temp > 0.0)
         c.set_temperature(average_temp);
       }
     }
   }
 }
 
 void OpenmcNekDriver::update_density()
+{
+  // Must be done in this order. Hence, update_density is public, whereas
+  // update_elem_densities and update_cell_densities are private
+  update_elem_densities();
+  update_cell_densities();
+}
+
+void OpenmcNekDriver::update_elem_densities()
 {
   if (nek_driver_->active()) {
+    // On Nek's master rank, d gets global data. On Nek's other ranks, d is empty.
+    auto d = nek_driver_->density();
+    // Update elem_densities_ on Nek's master rank only.
+    if (nek_driver_->comm_.rank == 0) {
+      elem_densities_ = d;
+    }
+    // Since OpenMC's and Nek's master ranks are the same, we know that elem_densities_ on
+    // OpenMC's master rank were updated.  Now we broadcast to the other OpenMC ranks.
+    // TODO: This won't work if the Nek/OpenMC communicators are disjoint
     if (openmc_driver_->active()) {
-      for (const auto& c : openmc_driver_->cells_) {
-        // Get corresponding global elements
-        const auto& global_elems = mat_to_elems_.at(c.material_index_);
-
-        bool any_in_fluid = false;
-        for (int elem : global_elems) {
-          if (elem_is_in_fluid_[elem] == 1) {
-            any_in_fluid = true;
-            break;
-          }
-        }
+      openmc_driver_->comm_.Bcast(elem_densities_.data(), n_global_elem_, MPI_DOUBLE);
+    }
+  }
+}
 
+void OpenmcNekDriver::update_cell_densities()
+{
+  if (nek_driver_->active() && openmc_driver_->active()) {
+    // Update cell densities for fluid cells.  Use cell_fluid_mask_ to do this
+    // TODO:  Might be able to use xtensor masking to do some of this
+    for (int i = 0; i < openmc_driver_->cells_.size(); ++i) {
+      if (cell_fluid_mask_[i] == 1) {
+        auto& c = openmc_driver_->cells_[i];
         double average_density = 0.0;
         double total_vol = 0.0;
-        for (int elem : global_elems) {
-          double T = temperatures_[elem];
-          double V = elem_volumes_[elem];
-
-          if (any_in_fluid) {
-            // nu1 returns specific volume in [m^3/kg]
-            double density = 1.0e-3 / iapws::nu1(pressure_, T);
-            average_density += density * V;
-            total_vol += V;
-          }
+        for (int e : mat_to_elems_.at(c.material_index_)) {
+          average_density += elem_densities_[e] * elem_volumes_[e];
+          total_vol += elem_volumes_[e];
         }
-
-        // Set density for cell instance
-        if (any_in_fluid)
-          c.set_density(average_density / total_vol);
+        double density = average_density / total_vol;
+        Expects(density > 0.0)
```

Comment:
Not sure if you want `Expects` or `Ensures` here.

Reply:
Again, `Expects` is fine. Add a semicolon to be safe